File: frontend-v2/src/features/chat/lib/compaction-view.ts
import type { CompactionPart, MessageWithParts } from "@/shared/types/api"

export interface CompactionView {
  id: string
  status: "running" | "completed" | "failed" | "interrupted"
  summary: string
}

export function compactionPart(message: MessageWithParts): CompactionPart | undefined {
  return message.parts.find((part): part is CompactionPart => part.type === "compaction")
}

export function isCompactionRequest(message: MessageWithParts): boolean {
  return message.role === "user" && (message.agent === "compaction" || Boolean(compactionPart(message)))
}

export function isCompactionMessage(message: MessageWithParts): boolean {
  return isCompactionRequest(message) || (message.role === "assistant" &&
    (message.agent === "compaction" || message.summary === true))
}

/** One process item per request, stable across deltas, settlement and paging.
 *  `agent` identifies the summary before its first token; `summary` alone is
 *  too late because the backend sets it only when the attempt settles. */
export function buildCompactionViews(messages: MessageWithParts[], streaming: boolean): CompactionView[] {
  const entries = new Map<string, { request?: MessageWithParts; summary?: MessageWithParts }>()
  for (const message of messages) {
    if (!isCompactionMessage(message)) continue
    const request = isCompactionRequest(message)
    const id = request ? message.id : (message.parent_id ?? message.id)
    entries.set(id, { ...entries.get(id), [request ? "request" : "summary"]: message })
  }
  return [...entries].map(([id, entry]) => {
    const descriptor = entry.request ? compactionPart(entry.request) : undefined
    const attempt = entry.summary
    const text = attempt?.parts
      .filter((part) => part.type === "text")
      .map((part) => part.text)
      .join("") || descriptor?.summary || ""
    const failed = Boolean(attempt?.error) || attempt?.finish === "error"
    const completed = !failed && (attempt?.finish === "stop" ||
      Boolean(descriptor?.replacement_id) || Boolean(descriptor?.summary))
    const last = attempt ?? entry.request
    const live = streaming && last?.id === messages.at(-1)?.id && !attempt?.finish
    return {
      id,
      summary: text,
      status: failed ? "failed" : completed ? "completed" : live ? "running" : "interrupted",
    }
  })
}

File: mobile/lib/features/chat/utils/compaction_view.dart
import '../../../shared/models/message.dart';
import '../../../shared/models/message_part.dart';

enum CompactionStatus { running, completed, failed, interrupted }

class CompactionView {
  const CompactionView({
    required this.id,
    required this.status,
    required this.summary,
  });

  final String id;
  final CompactionStatus status;
  final String summary;
}

CompactionPart? compactionPart(ChatMessage message) =>
    message.parts.whereType<CompactionPart>().firstOrNull;

bool isCompactionRequest(ChatMessage message) =>
    message.isUser &&
    (message.agent == 'compaction' || compactionPart(message) != null);

// The agent identifies streaming summaries before summary=true is persisted.
bool isCompactionMessage(ChatMessage message) =>
    isCompactionRequest(message) ||
    (message.isAssistant &&
        (message.agent == 'compaction' || message.summary == true));

/// Stable request identities survive streaming, history paging and retries.
List<CompactionView> buildCompactionViews(
  List<ChatMessage> messages,
  bool streaming,
) {
  final requests = <String, ChatMessage>{};
  final summaries = <String, ChatMessage>{};
  final ids = <String>{};
  for (final message in messages.where(isCompactionMessage)) {
    final request = isCompactionRequest(message);
    final id = request ? message.id : (message.parentId ?? message.id);
    ids.add(id);
    (request ? requests : summaries)[id] = message;
  }
  return [
    for (final id in ids)
      _view(id, requests[id], summaries[id], messages.lastOrNull, streaming),
  ];
}

CompactionView _view(
  String id,
  ChatMessage? request,
  ChatMessage? attempt,
  ChatMessage? last,
  bool streaming,
) {
  final descriptor = request == null ? null : compactionPart(request);
  final text = attempt?.parts.whereType<TextPart>().map((p) => p.text).join();
  final failed = attempt?.error != null || attempt?.finish == 'error';
  final completed =
      !failed &&
      (attempt?.finish == 'stop' ||
          (descriptor?.replacementId?.isNotEmpty ?? false) ||
          (descriptor?.summary?.isNotEmpty ?? false));
  final live =
      streaming &&
      (attempt ?? request)?.id == last?.id &&
      attempt?.finish == null;
  return CompactionView(
    id: id,
    summary: text != null && text.isNotEmpty ? text : descriptor?.summary ?? '',
    status: failed
        ? CompactionStatus.failed
        : completed
        ? CompactionStatus.completed
        : live
        ? CompactionStatus.running
        : CompactionStatus.interrupted,
  );
}
